### PR TITLE
[luci-interpreter] Add support for sub(tract) op

### PR DESCRIPTION
**Pull request by a-zh**
This commit adds `sub` operator in luci-interpreter

ONE-DCO-1.0-Signed-off-by: Anton Zhukov <anton.v.zhukov@gmail.com>

### CODE REVIEW

File: compiler/luci-interpreter/src/kernels/Sub.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+  if (need_broadcast)
+  {
+    tflite::reference_ops::BroadcastSubSlow(
+        params, getTensorShape(input1()), getTensorData<float>(input1()), getTensorShape(input2()),
+        getTensorData<float>(input2()), getTensorShape(output()), getTensorData<float>(output()));
+  }
+  else
+  {
+    tflite::reference_ops::Sub(params, getTensorShape(input1()), getTensorData<float>(input1()),
```
Comment (binarman) [inline compiler/luci-interpreter/src/kernels/Sub.cpp:82]: Not sure, maybe use optimized kernels here?
@s-barannikov @struss  what is our position about reference/optimized kernels?
I see it is mixed in other operations.

**Review by binarman — APPROVED** [review]
LGTM
Also see optional comment about optimized kernels

File: compiler/luci-interpreter/src/kernels/Sub.cpp (hunk 1 above)
Comment (s-barannikov) [reply compiler/luci-interpreter/src/kernels/Sub.cpp:82]: It is better to use optimized kernels, of course. Sometimes it is not that easy (see Conv2D for an example).
Some other ops may still use reference kernels due to historical reasons :)

File: compiler/luci-interpreter/src/kernels/CMakeLists.txt [hunk 2]
```diff
@@ -64,6 +64,8 @@ set(SOURCES
     Reverse.cpp
     Rsqrt.h
     Rsqrt.cpp
+    Sub.cpp
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/CMakeLists.txt:67]: Please keep alphabetical order (in other places too).
It is OK to fix incorrect ordering if you find one.

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+ * limitations under the License.
+ */
+
+#include "kernels/Sub.h"
+#include "kernels/TestUtils.h"
+
+#include <iostream>
+#include <fstream>
+#include <algorithm>
+#include <functional>
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:24]:
```suggestion
#include <algorithm>
```

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    kernel.execute();
+
+    EXPECT_THAT(dequantizeTensorData(output_tensor),
+                FloatArrayNear(output_data[i], kQuantizedTolerance));
+    EXPECT_THAT(extractTensorShape(output_tensor), ::testing::ElementsAreArray(output_shapes[i]));
+  }
+
+  // Inversion step for output_data, because subtract is not commutative operation
+  auto multiply = [](vector<float> &i) {
+    float invers{-1.0};
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:91]: Why don't just inline it?
NB I'd use simple range-based for loop.

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+TEST(SubTest, Uint8)
+{
+  initializer_list<int32_t> base_shape = {2, 3, 1, 2};
+  initializer_list<float> base_data = {-0.3f, 2.3f, 0.9f,  0.5f, 0.8f, -1.1f,
+                                       1.2f,  2.8f, -1.6f, 0.0f, 0.7f, -2.2f};
+  initializer_list<int32_t> test_shapes[] = {
+      {1, 1, 3, 2}, {1, 3, 1, 2}, {2, 1, 3, 1}, {2, 3, 1, 1}};
+  initializer_list<float> test_data = {0.2f, 0.3f, -0.4f, 0.5f, 1.0f, 0.9f};
+  initializer_list<int32_t> output_shapes[] = {
+      {2, 3, 3, 2}, {2, 3, 1, 2}, {2, 3, 3, 2}, {2, 3, 1, 2}};
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:55]: Please use `vector`s. It is very uncommon to use `initializer_list` for local variables (in fact, I've never seen this).

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    Tensor output_tensor = makeOutputTensor(DataType::FLOAT32);
+
+    SubParams params{};
+    params.activation = Activation::RELU;
+
+    Sub kernel(&input1_tensor, &input2_tensor, &output_tensor, params);
+    kernel.configure();
+    kernel.execute();
+
+    EXPECT_THAT(extractTensorData<float>(output_tensor), FloatArrayNear(test_outputs[i], 0.0001f))
```
Comment (s-barannikov) [inline compiler/luci-interpreter/src/kernels/Sub.test.cpp:150]: Should also check shapes. For that, you will need to change the type of `test_shapes` to `vector<vector<int32_t>>`.

File: compiler/luci-interpreter/src/kernels/CMakeLists.txt (hunk 2 above)
Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/CMakeLists.txt:67]: Done

Comment (s-barannikov) [reply compiler/luci-interpreter/src/kernels/CMakeLists.txt:67]: I mean, Squeeze < Sub < Tanh

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp (hunk 6 above)
Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/Sub.test.cpp:147]: Done

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp (hunk 5 above)
Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/Sub.test.cpp:52]: Done

File: compiler/luci-interpreter/src/kernels/Sub.cpp (hunk 1 above)
Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/Sub.cpp]: Done

File: compiler/luci-interpreter/src/kernels/CMakeLists.txt (hunk 2 above)
Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/CMakeLists.txt]: Yes of course.

File: compiler/luci-interpreter/src/kernels/Sub.test.cpp (hunk 4 above)
Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/Sub.test.cpp]: Constant is inlined.

File: compiler/luci-interpreter/src/kernels/Sub.cpp [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+{
+
+Sub::Sub(const Tensor *input1, const Tensor *input2, Tensor *output, const SubParams &params)
+    : KernelWithParams<SubParams>({input1, input2}, {output}, params)
+{
+}
+
+void Sub::configure()
+{
+  if (input1()->element_type() != input2()->element_type())
```
Comment (struss) [inline compiler/luci-interpreter/src/kernels/Sub.cpp:37]: What about using `LUCI_INTERPRETER_CHECK` on this?

Comment (a-zh) [reply compiler/luci-interpreter/src/kernels/Sub.cpp]: Ok. Done

**Review by s-barannikov — APPROVED** [review]
LGTM

File: compiler/luci-interpreter/src/kernels/Sub.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  calculateActivationRangeQuantized(_params.activation, output(), &activation_min, &activation_max);
+
+  tflite::ArithmeticParams params{};
+  params.left_shift = left_shift;
+  // The kernel expects inputs' zero points to be negated.
+  params.input1_offset = -input1()->zero_point(); // Note the '-'.
+  params.input1_multiplier = input1_multiplier;
+  params.input1_shift = input1_shift;
+  params.input2_offset = -input2()->zero_point(); // Note the '-'.
+  params.input2_multiplier = input2_multiplier;
```
Comment (jinevening) [inline compiler/luci-interpreter/src/kernels/Sub.cpp:112]: `input2_multiplier` is negated in TFLite source code (`Prepare8BitSubOp`)

**Review by jinevening — APPROVED** [review]
LGTM

**Review by struss — APPROVED** [review]
LGTM

**Review by binarman — APPROVED** [review]

**Pull request merged**